File: scripts/fix_headings.py
import os
import re
import json
import urllib.request

def load_env():
    env_path = os.path.join(os.path.dirname(__file__), '..', '.env.local')
    if os.path.exists(env_path):
        with open(env_path, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith('#'):
                    continue
                parts = line.split('=', 1)
                if len(parts) == 2:
                    key = parts[0].strip()
                    val = parts[1].strip().strip('"').strip("'")
                    os.environ[key] = val

load_env()

def main():
    db_url = os.environ.get('TURSO_CONNECTION_URL', '')
    db_token = os.environ.get('TURSO_AUTH_TOKEN', '')

    if not db_url:
        print('Error: TURSO_CONNECTION_URL is missing')
        return

    # Convert libsql:// to https:// for urllib
    http_url = db_url.replace('libsql://', 'https://')
    pipeline_url = f"{http_url}/v2/pipeline"

    # SQLite replace SQL statement
    sql_query = "UPDATE articles SET content = replace(replace(content, '<h2>## ', '<h2>'), '<h3>### ', '<h3>') WHERE slug = 'isro-ursc-research-internship-2026-complete-guide-eligibility'"

    payload = {
        "requests": [
            {
                "type": "execute",
                "stmt": {
                    "sql": sql_query
                }
            },
            {
                "type": "close"
            }
        ]
    }

    headers = {
        "Authorization": f"Bearer {db_token}",
        "Content-Type": "application/json"
    }

    try:
        req = urllib.request.Request(
            pipeline_url,
            data=json.dumps(payload).encode('utf-8'),
            headers=headers,
            method='POST'
        )
        with urllib.request.urlopen(req) as response:
            res_body = response.read().decode('utf-8')
            print('Successfully executed HTTP REST update pipeline!')
            print(res_body)
    except Exception as e:
        print(f"Error executing HTTP request: {e}")

if __name__ == '__main__':
    main()

File: scripts/fix_headings.js
const fs = require('fs');
const path = require('path');
const { createClient } = require('@libsql/client');

function loadEnv() {
  const envPath = path.join(__dirname, '..', '.env.local');
  if (fs.existsSync(envPath)) {
    const envContent = fs.readFileSync(envPath, 'utf8');
    envContent.split('\n').forEach(line => {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) return;
      const parts = trimmed.split('=');
      if (parts.length >= 2) {
        const key = parts[0]?.trim();
        const value = parts.slice(1).join('=').trim().replace(/(^["']|["']$)/g, '');
        if (key) process.env[key] = value;
      }
    });
  }
}

loadEnv();

async function main() {
  const logPath = path.join(__dirname, '..', 'fetch_output.txt');
  let log = '\n\n--- Database Heading Cleanup Log ---\n';

  const dbUrl = process.env.TURSO_CONNECTION_URL;
  const dbToken = process.env.TURSO_AUTH_TOKEN;

  if (!dbUrl) {
    log += 'Error: TURSO_CONNECTION_URL is missing\n';
    fs.appendFileSync(logPath, log, 'utf8');
    process.exit(1);
  }

  const client = createClient({ url: dbUrl, authToken: dbToken });
  try {
    const slug = 'isro-ursc-research-internship-2026-complete-guide-eligibility';
    const result = await client.execute({
      sql: 'SELECT id, content FROM articles WHERE slug = ?',
      args: [slug]
    });

    if (result.rows.length === 0) {
      log += 'Article not found in database\n';
    } else {
      let content = result.rows[0].content;
      log += 'Original content snippet: ' + content.substring(0, 150) + '\n';
      
      const original = content;
      // Replace markdown symbols inside HTML tags
      content = content.replace(/<h2>## /g, '<h2>');
      content = content.replace(/<h3>### /g, '<h3>');
      
      if (original !== content) {
        await client.execute({
          sql: 'UPDATE articles SET content = ? WHERE slug = ?',
          args: [content, slug]
        });
        log += 'Success: Heading hashes cleaned and saved to database!\n';
      } else {
        log += 'No changes needed, hashes already cleaned.\n';
      }
    }
  } catch (err) {
    log += 'Database Error: ' + err.message + '\n' + err.stack + '\n';
  } finally {
    client.close();
    fs.appendFileSync(logPath, log, 'utf8');
  }
}

main();
